### PR TITLE
Add 2x5 shower-shape and saturation info for electrons and photons

### PR DESCRIPTION
**Pull request by rafaellopesdesa**
I am working to update #17101 , but I would like to have all showershapes used in the regression already calculated inside the objects. We recently decided to sync the electron and photon regressions and the electron object did not have the 2x5 shower shapes, which are being added. I also added the saturation information that is now added as input for the regression (in both GsfElectron and Photon. This is a smaller big change, because it has to be propagated to miniAOD, which is also done here.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by rafaellopesdesa** [comment]
Apparently the part of the code that fills the saturation info for photons is crashing. I didn't see that before, I apologize. I will run these RelVals locally and update with the appropriate fix asap.

File: DataFormats/EgammaCandidates/src/classes_def.xml [hunk 1]
```diff
@@ -220,8 +225,12 @@
   <class name="reco::GsfElectron::PixelMatchVariables" ClassVersion="2">
     <version ClassVersion="2" checksum="3522539867"/>
   </class>
+  <class name="reco::GsfElectron::SaturationInfo" ClassVersion="0">
+   <version ClassVersion="0" checksum="3868678681"/>
```
Comment (slava77) [inline DataFormats/EgammaCandidates/src/classes_def.xml]: version=0 is a special number.
start from 3

**Comment by rafaellopesdesa** [comment]
@slava77  Hi Slava, thanks a lot for the help with versions. I didn't know about this detail. The code is updated.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by rafaellopesdesa** [comment]
I would like to use these new data members in the update of the PR for 90X regression... can I use it already?

**Comment by slava77** [comment]
On 2/7/17 9:57 AM, Rafael Lopes de Sa wrote:
> I would like to use these new data members in the update of the PR for
> 90X regression... can I use it already?
> 

Do you mean if you can combine this PR with the regression PR?
Sure, why not.

**Comment by rafaellopesdesa** [comment]
I was wondering what would be easier. I have a meeting tomorrow with AlCA to request the new payloads in the global tag so that the regression passes validation (they requested a more detailed explanation that was easier to do talking). I was thinking to then, tomorrow, submit the update to that PR. If it is easier, I merge this one together myself...

But I can also wait for this to be merged in the cmssw repository, since this one already passed all the tests.

File: DataFormats/EgammaCandidates/src/GsfElectron.cc [hunk 2]
```diff
@@ -71,6 +71,29 @@ GsfElectron::GsfElectron
  }
 
 GsfElectron::GsfElectron
+ ( int charge, const ChargeInfo & chargeInfo,
+   const GsfElectronCoreRef & core,
+   const TrackClusterMatching & tcm, const TrackExtrapolations & te,
+   const ClosestCtfTrack & ctfInfo,
```
Comment (slava77) [inline DataFormats/EgammaCandidates/src/GsfElectron.cc]: ctfInfo appears unnecessary.
Please remove it from this constructor.
Older constructors kept it to preserve interface backward compatibility, which is not applicable here.

Comment (slava77) [inline DataFormats/EgammaCandidates/src/GsfElectron.cc]: .. actually, do you even need to define this new constructor?
Can the old one be extended with one more argument?

**Comment by rafaellopesdesa** [comment]
Yes, it can. I was trying to preserve the "logic" of the code (well, more
like style than logic). Should I just extend the latest and more complete
constructor?

-- Rafael.

**Comment by slava77** [comment]
@rafaellopesdesa 
yes please,
just extend the latest most complete constructor (the one just above the proposed one)

**Comment by rafaellopesdesa** [comment]
OK, will do later today. Thanks.

**Comment by slava77** [comment]
@rafaellopesdesa 
should I expect an update here soon.
Please clarify.
Thank you.

**Comment by rafaellopesdesa** [comment]
Yes, I was in a conference these last days. Will update shortly.

Rafael

**Comment by slava77** [comment]
this needs a rebase
https://twiki.cern.ch/twiki/bin/viewauth/CMS/SWGuideHATS2016Git#CMSSW_rebasing_recipe

**Comment by slava77** [comment]
@rafaellopesdesa 
for a cleaner rebase, you will need to go back to the branch state before the merge ( 	de61ce5) and then follow the rebase instructions.

**Comment by rafaellopesdesa** [comment]
I will close and do together with the regression... it's done, I will submit now.

**Pull request closed without merging**